units/historical_test_data.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	j.SetScopes([]string{
+		j.Info.Project,
+		j.Info.Variant,
+		taskName,
+		j.Info.TestName,
+		j.Info.RequestType,
+		j.Info.Date.Format(model.HistoricalTestDataDateFormat),
+	})
+
+	return j
+}
+func makeHistoricalTestDataJob() *historicalTestDataJob {
+	j := &historicalTestDataJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    historicalTestDataJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+	return j
+}
+
+// getTaskName returns the task name to attribute to the. The display task name
+// is always prioritized over the execution task name, if the task is part of a
+// display task.
+func (j *historicalTestDataJob) getTaskName(info model.TestResultsInfo) string {
+	if info.DisplayTaskName != "" {
+		return info.DisplayTaskName
+	}
+	return info.TaskName
+}
+
+func (j *historicalTestDataJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if j.env == nil {
+		j.env = cedar.GetEnvironment()
+	}
+
+	conf := model.NewCedarConfig(j.env)
+	if err := conf.Find(); err != nil {
+		j.AddError(errors.Wrap(err, "finding cedar configuration"))
+		return
+	}
+	if conf.Flags.DisableHistoricalTestData {
+		return
+	}
+
+	skip, err := j.shouldNoop(ctx, conf)
+	j.AddError(errors.Wrapf(err, "checking if job needs to run"))
+	if skip {
+		return
+	}
+
+	htd, err := model.CreateHistoricalTestData(j.Info, conf.Bucket.HistoricalTestDataBucketType)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "creating historical test data"))
+		return
+	}
+	htd.Setup(j.env)
+	if err := htd.Find(ctx); err != nil {
+		// If the key does not yet exist, the test data model will be
+		// invalidated by Find(), so we have to re-create it. Otherwise, the
+		// job will fail to add the new key into S3 later.
+		htd, err = model.CreateHistoricalTestData(j.Info, conf.Bucket.HistoricalTestDataBucketType)
+		if err != nil {
+			j.AddError(errors.Wrap(err, "creating historical test data"))
+			return
+		}
+		htd.Setup(j.env)
+	}
+
+	switch j.Result.Status {
+	case "pass":
+		htd.NumPass += 1
+		dur := j.Result.TestEndTime.Sub(j.Result.TestStartTime)
+		htd.AverageDuration = (time.Duration(len(htd.Durations))*htd.AverageDuration + dur) / time.Duration(len(htd.Durations)+1)
+		htd.Durations = append(htd.Durations, dur)
+	case "fail", "silentfail":
+		htd.NumFail += 1
+	}
+
+	if err := htd.Save(ctx); err != nil {
+		j.AddError(errors.Wrap(err, "saving updated historical test data"))
+		return
+	}
+}
+
+// shouldNoop checks if this job must run for this project and task, depending
+// on whether historical test data is disabled for this job's project and
+// whether this project skips tasks matching this job's task name.
+func (j *historicalTestDataJob) shouldNoop(ctx context.Context, conf *model.CedarConfig) (bool, error) {
+	settings, err := j.getProjectSettings(ctx, conf)
+	if err != nil {
+		return false, errors.Wrapf(err, "getting historical test data settings for project '%s'", j.Info.Project)
```

Comment:
If we rely on skip in the `Run` shouldn't this be set to `true`? Unless we want to continue on error in this case.

Reply:
I set this to false because I think if we can't get the project settings (e.g. network failures, Evergreen is down), in the worst case, we'll end up including test results in the historical stats on an excluded task or a disabled project. it seems like it's safer to still record the test data rather than potentially drop a data point that we actually want.